auth/authzn.proto

```diff
@@ -0,0 +1,179 @@
+// authentication policy - where they do not support more granular authorization,
+// whereas others may have both specified.
+message Policy {
+    oneof protocol {
+        // console provides the configuration for authentication of the local
+        // system login.
+        Console console = 1;
+        // ssh provides the configuration for authentication of logins to the
+        // system via SSH.
+        SSH ssh = 2;
+        // gnmi provides a local authentication mechanism for all RPCs within
+        // the gNMI service, along with an authorization policy which allows
+        // users to be restricted to:
+        //  - specific RPCs (e.g., Get, Set, Subscribe, Capabilities)
+        //  - specific restrictions based on path within each RPC.
+        gNMI gnmi = 3;
+        // gnoi provides a local authentication mechanism for each gNOI RPC
+        // within the suite of gNOI microservices whose definitions are
+        // defined in github.com/openconfig/gnoi. Within specific microservices
+        // RPC-based restrictions may be realised.
+        gNOI gnoi = 4;
+        // gribi provides a local authentication mechanism for the gRIBI service.
+        gRIBI gribi = 5;
+    }
+}
+
+message Console {
+    // Console provides a username/password based authentication mechanism
+    // for logging in to the system. 
+    //
+    // The username of the user permitted to log in to the system via the 
+    // console. If the username is specified as an
+    // empty string, the supplied authentication details should apply to the
+    // root/superuser/admin user of the system.
+    string username = 1;
+    // TODO(robjs): How should this password be supplied to the device?
+    // Presumably we do not want it in cleartext. We could re-use the same
+    // definition that is used within the OpenConfig models:
+    // http://ops.openconfig.net/branches/models/master/docs/openconfig-system.html#type-crypt-password-type
+    string password = 2;
+}
+
+message SSH {
+    // SSH provides a key-based login for users onto the system. Only users
+    // that are specified within the SSH policy are permitted to login to the
+    // system.
+    //
+    // The username of a user for whom SSH login is permitted. If the value
+    // of the username field is an empty string, this corresponds to the
+    // root/superuser/admin user of the system.
+    string username = 1;
+    // authorized_keys specifies a list of SSH keys that are permitted for
+    // the user - each of which would correspond to a line within the
+    // authorized_users file on an openssh system. The CR/LF should be omitted
+    // from the supplied key.
+    repeated string authorized_keys = 2;
+}
+
+message gRPCID {
+    // The identify of a gRPC client is defined by its SPIFFE ID (see 
+    // https://spiffe.io/docs/latest/spiffe/concepts/) which is provided within
+    // the SVID - in this case, the X.509 client certificate.
+    string id = 1;
```

Comment:
SPIFFE ID is intended to identify a workload.  As shown in https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE-ID.md#22-path, some deployment may choose to encode more details in the SPIFFE ID Path components than what this authorization policy may care about.  As such, I don't think we can expect use of exact match with this "id" but allow inclusion of wildcards in the string.

Reply:
(disclaimer: I'm not super-familiar with how folks are intending to use SPIFFE IDs in production)

From what you're saying - and the doc - the intention here is that we might have `spiffe://some.domain.tld/network_services/some_service/some_compute_cluster` where we want to express this policy in terms of a subset of this path (e.g.,`spiffe://some.domain.tld/network_services/some_service/*`, or even `spiffe://*.domain.tld/network_services/some_service/some_compute_cluster`). Do we solely expect that this would be wildcarding entire path components (I presume it is not going to need to be regexp based for example)?

Reply:
Given spiffe-id doesn't have a very strict uri specification, I am afraid there need to be at least a way for users to specify multiple wildcards if we don't want to go as far as supporting regexp.  I think the owner identity will be the most interesting to most use cases.  Extending the "identifying service owners" example there, I can see the "ns" and "sa" kubernetes specific markers can be replaced with other such as "user-group" and "user":
spiffe://production.some_compute_cluster.example.com/group/foo/user/bar
or simply 
spiffe://production.some_compute_cluster.example.com/foo/bar

In this case, for matching all users of group Foo and regardless of which cluster the job is running from, we will want to allow users to specify:
spiffe://production.*.example.com/group/foo/user/*
or
spiffe://production.*.example.com/foo/*

What do you think?